torch/csrc/autograd/python_function.cpp

```diff
@@ -0,0 +1,1083 @@
+#include "torch/csrc/autograd/python_function.h"
+
+#include <Python.h>
+#include <structmember.h>
+#include <unordered_map>
+#include <unordered_set>
+#include <exception>
+#include <THPP/THPP.h>
+
+#include "THP.h"
+#include "torch/csrc/autograd/python_native_function.h"
+#include "torch/csrc/DynamicTypes.h"
+#include "torch/csrc/utils/auto_gil.h"
+
+#ifdef WITH_CUDA
+#include "cuda/AutoGPU.h"
+#endif
+
+using namespace torch;
+using namespace torch::autograd;
+using thpp::Tensor;
+
+PyObject *THPFunctionClass = NULL;
+PyObject *THPStochasticFunctionClass = NULL;
+
+
+// Throwing this exception means that the python error flags have been already
+// set and control should be immediately returned to the interpreter.
+
+#define THPFunction_assert(condition, ...)                                     \
+  if (!(condition)) { THPUtils_setError(__VA_ARGS__); throw python_error(); }
+
+
+namespace torch { namespace autograd {
+
+auto PyFunction::apply(const variable_list& gradOutputs) -> variable_list {
+  AutoGIL gil;
+
+  THPObjectPtr pyGradOutputs = PyTuple_New(gradOutputs.size());
+  if (!pyGradOutputs) throw python_error();
+
+  for (size_t i = 0; i != gradOutputs.size(); ++i) {
+    PyObject* gradOutput;
+    if (gradOutputs[i]) {
+      gradOutput = createPyObject(*gradOutputs[i]->data);
+      if (!gradOutput) throw python_error();
+    } else {
+      gradOutput = Py_None;
+      Py_INCREF(gradOutput);
+    }
+    PyTuple_SET_ITEM(pyGradOutputs.get(), i, gradOutput);
+  }
+
+  THPObjectPtr r = PyObject_CallMethod(
+      obj, "_do_backward", "OO", pyGradOutputs.get(), Py_True);
+  if (!r) throw python_error();
+
+  auto num_outputs = PyTuple_GET_SIZE(r.get());
+  variable_list results(num_outputs);
+  for (int i = 0; i != num_outputs; ++i) {
+    PyObject* obj = PyTuple_GET_ITEM(r.get(), i);
+    if (obj != Py_None) {
+      if (!THPModule_isTensor(obj)) {
+        throw std::runtime_error(std::string("expected Tensor (got '") + Py_TYPE(obj)->tp_name + "')'");
+      }
+      results[i] = std::make_shared<Variable>(createTensor(obj), 0, 1);
+    }
+  }
+  return results;
+}
+
+auto PyFunction::releaseVariables() -> void {
+  auto f = (THPFunction*) obj;
+  delete f->saved_variables;
```

Comment:
Good catch. I'm not releasing the GIL in engine for now, until I make sure we re-acquire it in all the right places.